MRG: Reference compute_epochs_csd in other spectral estimation docs.

nfoti opened the pull request:
Added references to `compute_epochs_csd` function to the documentation of other spectral estimation functions.

Eric89GXL commented:
LGTM

I wonder if we should rename it to be more consistent with the `psd_*` functions at some point

nfoti commented:
That could be helpful. Even going as far as `cross_spectrum_epochs` may be helpful as I don't know how often `csd` is actually used. In statistics we usually just call it the 'spectral density matrix' or less commonly the 'cross-spectrum'.

nfoti commented:
To do for this PR:

[] Refactor `compute_csd_epochs` to also have a function that works with `ndarrays`.
[] Rename `compute_epochs_csd` to something like `csd_multitaper` to be consistent.

dengemann commented:
> Refactor compute_csd_epochs to also have a function that works with ndarrays.

Is this really needed? Can we have a separate function for this? Or simply providing scipy like API, ultimately pointing people to scipy....

Are we really sure about this renaming business? We have changed names here quite a lot. This has not necessarily improved things. It generates frustrated users instead.

Eric89GXL commented:
> Are we really sure about this renaming business? We have changed names here quite a lot.

It would be to be more consistent with the `psd_*` functions, so I assume it would be as worth it as that was. But we don't need to do it I suppose. At the very least the new functions could be made consistent.

dengemann commented:
We can do it just let's make sure that we don't refactor only to refactor
soon again.

On Tue, Jun 28, 2016 at 10:30 PM Eric Larson <notifications@github.com>
wrote:

> Are we really sure about this renaming business? We have changed names
> here quite a lot.
>
> It would be to be more consistent with the psd_* functions, so I assume
> it would be as worth it as that was. But we don't need to do it I suppose.
> At the very least the new functions could be made consistent.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/3355#issuecomment-229173394>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AB0fil2JbbNkXSXFW1dmQ3eS8su0eiAWks5qQYSCgaJpZM4JAbNp>
> .
>

nfoti commented:
The code I just pushed contains a new function `csd_array` (the name can change) that implements the current `compute_epochs_csd` function on an `ndarray` (and adapted the current tests for it too). At first I tried to make a function that `compute_epochs_csd` could call, but that turned out to be kind of ugly. The unfortunate part about this is that there are two functions that do the same thing on two different data structures, but at least now there is an interface to compute cross spectral-densities on `ndarrays`.

agramfort commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +255,162 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+def csd_array(X, sfreq, mode='multitaper', fmin=0, fmax=np.inf,
+              fsum=True, n_fft=None, mt_bandwidth=None,
+              mt_adaptive=False, mt_low_bias=True, verbose=None):
+    """Estimate cross-spectral density from n_trials x n_series x n_times
```
n_trials x n_series x n_times
->
n_epochs x n_channels x n_times

agramfort replied:
@nfoti agree on this?

Eric89GXL replied:
"from an array" or so only, give the shape in the parameter list instead

Eric89GXL replied:
This still needs to be resolved

Eric89GXL replied:
This still needs to be resolved

nfoti replied:
I'm addressing this now. For this function I personally like the shapes called something like `n_replicates x n_series x n_times` since the function is not restricted to operating on `Epochs`. What do you guys think?

agramfort commented:
travis is not happy

regarding code dupe. Can you reuse the csd_array function in compute_epochs_csd ?

nfoti commented:
I have fixed the travis errors and have also removed a lot of duplicate code by pushing most of the work to helper functions that both functions can compute. It wouldn't be too hard to create additional versions of these functions like `csd_source_estimate` if they ended up being useful, but as it is, it's really helpful to have a function to compute a multitaper estimate of the cross-spectrum from an `ndarray`.

I also renamed `compute_epochs_csd` to `csd_epochs` everywhere (according to grep) to make it consistent with the other functions for spectral densities. I also added a deprecation warning to `compute_epochs_csd`.

agramfort commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +220,201 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+            n_trials = 1
+            n_series, n_times = X.shape
+        else:
+            err_str = "X must be either n_trials x n_seris x n_times "
```
n_seris

agramfort commented on mne/time_frequency/tests/test_csd.py:
```diff
@@ -161,3 +159,188 @@ def test_compute_epochs_csd_on_artificial_data():
+                                ref_meg=False, exclude='bads')
+
+    epochs_data = [e[picks_meeg][:, tslice] for e in epochs]
+    X = _stack(epochs_data, axis=0)
```
how about just using np.concatenate(epochs_data, axis=0) and avoid the stack in fixes?

agramfort commented on mne/time_frequency/tests/test_csd.py:
```diff
@@ -161,3 +159,188 @@ def test_compute_epochs_csd_on_artificial_data():
+    assert_array_equal(csd_fsum, csd_sum)
+
+
+def test_csd_on_artificial_data():
```
how long do the tests take to run?

agramfort commented:
that's it for me.

also travis is not happy

Eric89GXL commented on mne/time_frequency/csd.py:
```diff
@@ -48,11 +50,23 @@ def __repr__(self):
         return '<CrossSpectralDensity  |  %s>' % s
 
 
+@deprecated()
```
you should add a message like:
```
@deprecated('compute_epochs_csd has been deprecated and will be removed in 0.14, use csd_epochs instead')
```

Eric89GXL commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +220,201 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+    """Estimate cross-spectral density from n_trials x n_series x n_times
+       ndarray.
+
+    Note: Results are scaled by sampling frequency for compatibility with
```
to get a proper directive use:

```
.. note:: Results are scaled by the sampling frequency
          for compatibility with MATLAB.
```

Eric89GXL commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +220,201 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+
+    Parameters
+    ----------
+    X : numpy.ndarray or list of numpy.ndarray
```
maybe just `array-like, shape (...)`

Eric89GXL commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +220,201 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+        Spectrum estimation mode can be either: 'multitaper' or 'fourier'.
+    fmin : float
+        Minimum frequency of interest.
+    fmax : float | np.inf
```
`np.inf` is a float, no need to list it

Eric89GXL commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +220,201 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+        if len(shapes) != 1:
+            raise ValueError("All arrays in X must be the same shape")
+        n_trials = len(X)
+        n_series, n_times = X[0].shape
```
This all looks a bit complicated. How about instead:
```
X = np.asarray(X, dtype=float)  # this will ensure all sizes work
if X.ndim != 3:
    raise ValueError(...)
```
If a user wants to use a single trial they can expand axes with `np.newaxis`.

Eric89GXL commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +220,201 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+    Returns
+    -------
+    csd : numpy.ndarray
+        The computed cross spectral-density (either summed or not).
```
give possible shapes based on `fsum` value

Eric89GXL commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +220,201 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+    freq_mask_mt = freq_mask[orig_frequencies >= 0]
+
+    # Compute CSD for each trial
+    for ti in xrange(n_trials):
```
`xrange` does not exist on Py3k. `n_trails` will probably never be huge so just use `range` here. In the future if you really need it you can get it from `six` (which we keep a copy of in `externals`)

Eric89GXL commented:
Otherwise LGTM

nfoti commented:
I've addressed the points raised by @agramfort and @Eric89GXL.

@agramfort The tests take about 4 seconds total to run the 4 tests.

I will change this PR to [MRG] once all tests pass.

Eric89GXL commented:
... and in addition to changing the title to MRG you should also ping again for review, as we don't get GitHub notifications for title changes. Plus that way @agramfort won't have to worry about his inbox being empty.

Eric89GXL commented:
LGTM, the only thing I would ask is, can you make the tests shorter by using fewer time points, fewer epochs, and/or fewer channels? 4 sec doesn't seem like much but it can add up over all the PRs we get, but if it's the minimum you can get while safely checking the functions we can live with it.

agramfort commented:
ok I will review when set to MRG then

nfoti commented:
@Eric89GXL I got the total time taken by all the tests to be about 2.7 seconds from 4 seconds originally via obvious optimizations (each test only loads the data it needs and trying less combinations of redundant multi taper parameters were the major ones). To go quicker would require a lot more work I think. Hopefully this is a big enough speedup to merge these functions.

Eric89GXL commented on mne/time_frequency/csd.py:
```diff
@@ -155,33 +171,9 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
-        raise ValueError('Mode has an invalid value.')
+    pars = _compute_csd_params(n_times, sfreq, mode, mt_bandwidth, mt_low_bias,
+                               mt_adaptive)
+    window_fun, eigvals, n_tapers, mt_adaptive = pars
```
FYI to make this PEP8 but a little bit more compact you can do:
```
window_fun, eigvals, n_tapers, mt_adaptive = _compute_csd_params(
    n_times, sfreq, mode, mt_bandwidth, mt_low_bias, mt_adaptive)
```
...but don't really need to do it here

Eric89GXL commented on mne/time_frequency/tests/test_csd.py:
```diff
@@ -161,3 +163,163 @@ def test_compute_epochs_csd_on_artificial_data():
+
+def test_compute_csd():
+    """Test computing cross-spectral density from ndarray
+    """
```
Turns out this should all be on one line `"""Test ... """`. Feel free to change the other instances in the file if you want, which I assume is what led you astray.

Eric89GXL commented on mne/time_frequency/tests/test_csd.py:
```diff
@@ -161,3 +163,163 @@ def test_compute_epochs_csd_on_artificial_data():
+
+    csd_sum = np.sum(csds, axis=2)
+
+    assert(csds.shape[2] == 2)
```
use `assert_equal` from `nose.tools` or `numpy.testing`

Eric89GXL replied:
This still needs to be resolved

nfoti replied:
Addressed in new commit.

Eric89GXL commented on mne/time_frequency/tests/test_csd.py:
```diff
@@ -110,47 +112,47 @@ def test_compute_epochs_csd():
     # Power per sample should not depend on time window length
-    for tmax in [0.2, 0.4, 0.6, 0.8]:
+    for tmax in [0.2, 0.8]:
         for add_n_fft in [30, 0, 30]:
```
I see you use this elsewhere -- why repeat the 30 condition? Does something accumulate / change?

nfoti replied:
@Eric89GXL This is a relic from the original tests. I'll fix this to just do `0` and `30`.

Eric89GXL commented:
2.7 seconds is okay if you think it's reasonably optimized. Other than my last few comments LGTM

nfoti commented:
@Eric89GXL I addressed your comments. Let me know if there's anything else.

Eric89GXL commented:
Looks like you missed a couple of comments

Eric89GXL commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +221,187 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+
+    Parameters
+    ----------
+    X : array-like, shape (n_replicates, n_series, n_times)
```
This is true in the most general sense, yes. But 90% of the time (or more?) it will be `(n_trials, n_series, n_times)` for MNE-Python'ers, so I'd prefer that. At the very least you could expand below to say "The time series data, e.g. with shape (n_trials, n_channels, n_times) or (n_trials, n_src, n_times)". But wait to see what @agramfort says.

agramfort replied:
we should stick to standard names we use everywhere in the code base. Please at least rename n_series to n_channels

kingjr replied:
We don't usually use `n_trials` in MNE. Is it applicable to something else than epochs? else why not `n_epochs` ?

nfoti replied:
@Eric89GXL @agramfort Since this is the function that is generally applicable shouldn't we avoid epoch-specific language? I agree with Eric that providing explicit examples in the docstring would be helpful, but I think it's more confusing to use `n_channels` when this function could be used for source-space analysis where you don't have channels. By using `n_channels` we're catering to users who are analyzing `Epoch`s, but there is an explicit function for that analysis that uses epoch-specific terminology. Does this sound reasonable to you? If you really feel strongly about using the epoch-specific terminology I'll use it, but it does make me uncomfortable having the generic version of the function use vocabulary from a specific use-case.

Eric89GXL replied:
I'm okay with it either way

agramfort replied:
ok @nfoti but then please explicit in the docstring below what n_series etc. can be as you explained here.

nfoti commented:
@Eric89GXL I've addressed the outstanding comments in newest commits.

kingjr commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +221,187 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+    sfreq : float
+        Sampling frequency of observations.
+    mode : str
+        Spectrum estimation mode can be either: 'multitaper' or 'fourier'.
```
Defaults to 'multitaper'

nfoti replied:
@Eric89GXL @kingjr There does not seem to be a consistent style in mne-python about whether default values are in the docstring or are only in the function definition. I am a fan of not duplicating them in the docstring as the function definition is basically always printed with the documentation and it's just another thing to forget to change if the default value is ever changed.

kingjr replied:
ok

kingjr commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +221,187 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+    mode : str
+        Spectrum estimation mode can be either: 'multitaper' or 'fourier'.
+    fmin : float
+        Minimum frequency of interest.
```
Defaults to ..., same below

kingjr commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +221,187 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+
+    Returns
+    -------
+    csd : numpy.ndarray, shape (n_freqs, n_series, n_series) if fsum is True,
```
"numpy.ndarray" => array

kingjr replied:
same everywhere

kingjr commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +221,187 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+    logger.info('Computing cross-spectral density from ndarray...')
+    pars = _compute_csd_params(n_times, sfreq, mode, mt_bandwidth, mt_low_bias,
+                               mt_adaptive)
+    window_fun, eigvals, n_tapers, mt_adaptive = pars
```
unnecessary temp variable

kingjr commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +221,187 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+    freq_mask_mt = freq_mask[orig_frequencies >= 0]
+
+    # Compute CSD for each trial
+    for ti in range(n_replicates):
```
I think you can just do `for xi in X:`

kingjr commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +221,187 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+
+
+def _compute_csd_params(n_times, sfreq, mode, mt_bandwidth, mt_low_bias,
+                        mt_adaptive):
```
add some aux. docs

kingjr commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +221,187 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+    if mode == 'multitaper':
+        # Compute standardized half-bandwidth
+        if mt_bandwidth is not None:
+            half_nbw = float(mt_bandwidth) * n_times / (2 * sfreq)
```
if int, use // instead of /, else use 2. instead of 2

kingjr commented on mne/time_frequency/csd.py:
```diff
@@ -255,3 +221,187 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+                                       np.newaxis]
+        else:
+            # Hack so we can sum over axis=-2
+            weights = np.array([1.])[:, None, None, None]
```
use np.newaxis instead of None

kingjr commented on mne/time_frequency/tests/test_csd.py:
```diff
@@ -8,57 +8,60 @@
+        epochs = mne.Epochs(raw, events, event_id, tmin, tmax, proj=True,
+                            picks=picks, baseline=(None, 0), preload=True,
+                            reject=dict(grad=4000e-13, mag=4e-12))
+    elif mode == 'sin':
```
wdym sin? simulated?

kingjr replied:
my bad

kingjr commented on mne/time_frequency/tests/test_csd.py:
```diff
@@ -161,3 +161,161 @@ def test_compute_epochs_csd_on_artificial_data():
+    # Power per sample should not depend on time window length
+    for tmax in [0.2, 0.8]:
+        tend = np.where(epochs.times <= tmax)[0][-1] + 1
+        tslice = slice(None, tend, None)
```
doesn't `tslice = np.where(epochs.times <= tmax) work?

Eric89GXL replied:
that would work, but it wouldn't actually be a slice but an array of indices, and thus have different behavior (copy on access instead of view on access)

kingjr commented on mne/time_frequency/tests/test_csd.py:
```diff
@@ -161,3 +161,161 @@ def test_compute_epochs_csd_on_artificial_data():
+                                            fmin=0, fmax=np.inf, n_fft=n_fft)
+
+            fourier_power_per_sample = np.abs(data_csd_fourier[0, 0]) *\
+                sfreq / n_fft
```
first_samp = data_csd_fourier[0, 0]
fourier_power_per_sample = np.abs(first_samp) * sfreq / n_fft

kingjr commented on mne/time_frequency/tests/test_csd.py:
```diff
@@ -161,3 +161,161 @@ def test_compute_epochs_csd_on_artificial_data():
+            fourier_power_per_sample = np.abs(data_csd_fourier[0, 0]) *\
+                sfreq / n_fft
+            assert_true(abs(signal_power_per_sample -
+                            fourier_power_per_sample) < 0.003)
```
why 0.003? don't you want np.testing.array_almost_equal?

nfoti replied:
@kingjr This is what was present in the existing tests that I adapted for the new functions. I think questioning the `0.003` belongs in a separate issue since it was already present in the test prior to my edits.

kingjr replied:
ok

kingjr commented on mne/time_frequency/tests/test_csd.py:
```diff
@@ -161,3 +161,161 @@ def test_compute_epochs_csd_on_artificial_data():
+            assert_true(abs(signal_power_per_sample -
+                            fourier_power_per_sample) < 0.003)
+        # Power per sample should not depend on number of tapers
+        for n_tapers in [1, 2, 3, 5]:
```
1, 2, 5, should suffice?

kingjr commented on mne/time_frequency/tests/test_csd.py:
```diff
@@ -161,3 +161,161 @@ def test_compute_epochs_csd_on_artificial_data():
+                mt_bandwidth = sfreq / float(n_samples) * (n_tapers + 1)
+                data_csd_mt, _ = csd_array(epochs._data[:, :, tslice],
+                                           sfreq, mode='multitaper',
+                                           fmin=0, fmax=np.inf,
```
don't specify if default params

kingjr commented:
Do we need to call it `csd_epochs`? Why not just `csd`? Specifying the input type seems a bit redundant, from what I understand, the function could easily be adapted to evoked, and perhaps raw data?

kingjr commented:
Also, CSD is an acronym for 'current source density' too which is common in EEG analysis. I'm wondering whether it wouldn't be a good thing to make the name more explicit.

jona-sassenhagen commented:
>Also, CSD is an acronym for 'current source density' too which is common in EEG analysis. I'm wondering whether it wouldn't be a good thing to make the name more explicit.

+1

Eric89GXL commented:
It should probably stay `csd_epochs` to be consistent with existing `*psd_epochs` functions, though

agramfort commented:
CIs are not happy

nfoti commented:
Something subtle made `test_dics.py`. I can't run that test locally so am slowly working my way through the changes to figure out what it is.

nfoti commented:
@Eric89GXL @agramfort I've addressed the concerns from everyone above. Let me know if the documentation for `csd_array` is clear enough w.r.t. the shapes of the array.

The tests that Travis barfed on were for `fieldtrip_client` where the error `Could not connect to FieldTrip buffer` was thrown. I don't think this has anything to do with the code I pushed but let me know and I can run the tests again.

Eric89GXL commented:
And this:
```
Running flake8
mne/time_frequency/csd.py:342:1: W293 blank line contains whitespace
1
make: *** [flake] Error 1
```

agramfort commented:
@nfoti please update what's new about API change then +1 for merge.

nfoti commented:
@Eric89GXL @agramfort Added changes to API section of `whats_new.rst`. Let me know if you want it reworded or more/less details are required.

Eric89GXL commented on doc/whats_new.rst:
```diff
@@ -64,6 +64,7 @@ BUG
 
 API
 ~~~
+    - Adds new function :func:`mne.time_frequency.csd_array` to compute the cross-spectral density of multivariate signals stored in an array. Also renamed :func:`mne.time_frequency.csd_compute_epochs` to :func:`mne.time_frequency.csd_epochs` for consistency with other names. :func:`mne.time_frequency.csd_epochs` is now deprecated and will be removed in mne 0.14, by `Nick Foti`_
```
the adding of the new function goes under changelog, only the API change goes under here (the second part about renaming and deprecation)

nfoti replied:
@Eric89GXL Sounds good, fixed in c43bf91.

Eric89GXL replied:
It just occurred to me that you also need to add a URL at the bottom of the RST for your name to render properly as a link, see other examples

nfoti replied:
@Eric89GXL Cool, I added myself and pushed.

nfoti commented:
@Eric89GXL Rebased onto `origin/master` and did some squashing as there were a lot of unimportant commits.

nfoti commented:
The Travis failures do not appear to have anything to do with my code and I'm not sure why they are arising.

agramfort commented:
@Eric89GXL merge if you're happy

Eric89GXL commented:
@nfoti sorry for the pain but could you rebase again? It should fix the odd failures

nfoti commented:
@Eric89GXL No problem, just rebased and pushed. Will ping when it looks like it passes.

nfoti commented:
@Eric89GXL All tests passed.

Eric89GXL commented:
Thanks @nfoti

agramfort commented:
Thanks !






On Tue, Jul 19, 2016 at 2:47 AM +0200, "Eric Larson" <notifications@github.com> wrote:












Thanks @nfoti



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.
